File: node/be/api/venues/minicex.js
'use strict';

/**
 * HTTP client for the mini-cex service. Node's built-in fetch; no library.
 *
 * A response is returned whole -- status, lower-cased headers, parsed body --
 * so a step can assert on any of them. Only a transport failure is
 * ApiUnreachable (grades Blocked); a 4xx/5xx is an answer, often the one under
 * test.
 */

const BASE = (process.env.MINI_CEX_URL || 'http://127.0.0.1:8110').replace(/\/+$/, '');

class ApiUnreachable extends Error {}

class MiniCex {
  constructor(base = BASE) { this.base = base; }

  async request(method, path, { token, body, headers = {} } = {}) {
    const h = { ...headers };
    if (token) h.authorization = 'Bearer ' + token;
    const init = { method, headers: h };
    if (body !== undefined) { h['content-type'] = 'application/json'; init.body = JSON.stringify(body); }
    let res;
    try { res = await fetch(this.base + path, init); } catch (err) {
      throw new ApiUnreachable('mini-cex at ' + this.base + ' did not answer ' + method + ' ' + path + ': ' + (err.cause && err.cause.message ? err.cause.message : err.message));
    }
    const text = await res.text();
    let parsed = null; try { parsed = text ? JSON.parse(text) : null; } catch { parsed = null; }
    return { status: res.status, headers: Object.fromEntries([...res.headers.entries()].map(([k, v]) => [k.toLowerCase(), v])), body: parsed, text };
  }
  get(p, o) { return this.request('GET', p, o); }
  post(p, body, o = {}) { return this.request('POST', p, { ...o, body }); }

  /** Create a fresh account with a unique handle; returns { id, handle, token }. */
  async createAccount(prefix = 'acct') {
    const handle = prefix + '-' + Date.now().toString(36) + Math.random().toString(36).slice(2, 6);
    const r = await this.post('/accounts', { handle });
    if (r.status !== 201) throw new Error('create account failed: HTTP ' + r.status + ' ' + r.text);
    return r.body;
  }
  deposit(token, asset, amount, key) { return this.post('/deposit', { asset, amount, idempotency_key: key }, { token }); }
  withdraw(token, asset, amount, key) { return this.post('/withdraw', { asset, amount, idempotency_key: key }, { token }); }
  transfer(token, toHandle, asset, amount, key) { return this.post('/transfer', { to_handle: toHandle, asset, amount, idempotency_key: key }, { token }); }
  quote(fromAsset, toAsset, fromAmount) { return this.post('/quote', { from_asset: fromAsset, to_asset: toAsset, from_amount: fromAmount }); }
  swap(token, body) { return this.post('/swap', body, { token }); }
  bridgeWithdraw(token, body) { return this.post('/bridge/withdraw', body, { token }); }
  bridgeDeposit(body) { return this.post('/bridge/deposit', body); }
  bridgeConfirm(id) { return this.post('/bridge/' + id + '/confirm', undefined); }

  /** A funded account with `amount` base units of `asset` deposited. */
  async funded(asset, amount, prefix = 'acct') {
    const acct = await this.createAccount(prefix);
    const r = await this.deposit(acct.token, asset, amount);
    if (r.status !== 201) throw new Error('funding failed: HTTP ' + r.status + ' ' + r.text);
    return acct;
  }
}

module.exports = { MiniCex, ApiUnreachable, BASE };

File: python/be/api/venues/minicex.py
"""HTTP client for the mini-cex service. Mirror of node/be/api/venues/minicex.js.
urllib only. A transport failure is ApiUnreachable (grades Blocked); a 4xx/5xx
is an answer, often the one under test.
"""

from __future__ import annotations

import json
import os
import time
import urllib.error
import urllib.request

BASE = os.environ.get("MINI_CEX_URL", "http://127.0.0.1:8110").rstrip("/")


class ApiUnreachable(Exception):
    pass


class MiniCex:
    def __init__(self, base: str = BASE) -> None:
        self.base = base

    def request(self, method, path, token=None, body=None, headers=None):
        h = dict(headers or {})
        if token:
            h["Authorization"] = "Bearer " + token
        data = None
        if body is not None:
            h["Content-Type"] = "application/json"
            data = json.dumps(body).encode()
        req = urllib.request.Request(self.base + path, data=data, headers=h, method=method)
        try:
            with urllib.request.urlopen(req, timeout=15) as res:
                status, text = res.status, res.read().decode("utf-8", "replace")
        except urllib.error.HTTPError as err:
            status, text = err.code, err.read().decode("utf-8", "replace")
        except (urllib.error.URLError, TimeoutError, OSError) as err:
            raise ApiUnreachable(f"mini-cex at {self.base} did not answer {method} {path}: {err}") from err
        try:
            parsed = json.loads(text) if text else None
        except json.JSONDecodeError:
            parsed = None
        return {"status": status, "body": parsed, "text": text}

    def get(self, p, **kw):
        return self.request("GET", p, **kw)

    def post(self, p, body=None, **kw):
        return self.request("POST", p, body=body, **kw)

    def create_account(self, prefix="acct"):
        handle = f"{prefix}-{int(time.time()*1000):x}{os.urandom(2).hex()}"
        r = self.post("/accounts", {"handle": handle})
        if r["status"] != 201:
            raise RuntimeError("create account failed: HTTP " + str(r["status"]) + " " + r["text"])
        return r["body"]

    def deposit(self, token, asset, amount, key=None):
        return self.post("/deposit", {"asset": asset, "amount": amount, "idempotency_key": key}, token=token)

    def withdraw(self, token, asset, amount, key=None):
        return self.post("/withdraw", {"asset": asset, "amount": amount, "idempotency_key": key}, token=token)

    def transfer(self, token, to_handle, asset, amount, key=None):
        return self.post("/transfer", {"to_handle": to_handle, "asset": asset, "amount": amount, "idempotency_key": key}, token=token)

    def quote(self, from_asset, to_asset, from_amount):
        return self.post("/quote", {"from_asset": from_asset, "to_asset": to_asset, "from_amount": from_amount})

    def swap(self, token, body):
        return self.post("/swap", body, token=token)

    def bridge_withdraw(self, token, body):
        return self.post("/bridge/withdraw", body, token=token)

    def bridge_deposit(self, body):
        return self.post("/bridge/deposit", body)

    def bridge_confirm(self, op_id):
        return self.post(f"/bridge/{op_id}/confirm")

    def funded(self, asset, amount, prefix="acct"):
        acct = self.create_account(prefix)
        r = self.deposit(acct["token"], asset, amount)
        if r["status"] != 201:
            raise RuntimeError("funding failed: HTTP " + str(r["status"]) + " " + r["text"])
        return acct
